tests/test_args_specs.py

```diff
@@ -0,0 +1,311 @@
+    },
+    'raster': {'code', 'number', 'ratio'},
+    'vector': {
+        'code',
+        'freestyle_string',
+        'number',
+        'option_string',
+        'percent',
+        'ratio'},
+    'csv': {
+        'boolean',
+        'code',
+        'freestyle_string',
+        'number',
+        'option_string',
+        'percent',
+        'raster',
+        'ratio',
+        'vector'},
+    'directory': {'csv', 'directory', 'file', 'raster', 'vector'}
+}
+
+
+class ValidateArgsSpecs(unittest.TestCase):
+
+    def test_model_specs(self):
+
+        for model_name, val in cli._MODEL_UIS.items():
+            # val is a collections.namedtuple, fields accessible by name
+            model = importlib.import_module(val.pyname)
+
+            # validate that each arg meets the expected pattern
+            # save up errors to report at the end
+            for key, arg in model.ARGS_SPEC['args'].items():
+                # the top level should have 'name' and 'about' attrs
+                # but they aren't required at nested levels
+                self.validate(
+                    arg,
+                    f'{model_name}.{key}',
+                    valid_types=valid_nested_types['all'],
+                    required_attrs=['name', 'about'])
+
+    def validate(self, arg, name, valid_types, is_pattern=False, required_attrs=[]):
+        """
+        Recursively validate nested args against the ARGS_SPEC standard.
+
+        Args:
+            arg (dict): any nested arg component of an ARGS_SPEC
+            name (str): name to use in error messages to identify the arg
+            valid_types (list[str]): a list of the arg types that are valid
+                for this nested arg (due to its parent's type).
+            is_pattern (bool): if True, the arg is validated as a pattern (such
+                as for user-defined  CSV headers, vector fields, or directory
+                paths).
+            required_attrs (list[str]): a list of attributes that must be in
+                the arg dictionary regardless of type
+
+        Returns:
+            None
+
+        Raises:
+            AssertionError if the arg violates the standard
+        """
+        with self.subTest(nested_arg_name=name):
+            for attr in required_attrs:
+                self.assertTrue(attr in arg)
+
+            # arg['type'] can be either a string or a set of strings
+            types = arg['type'] if isinstance(
+                arg['type'], set) else [arg['type']]
+            attrs = set(arg.keys())
+            for t in types:
+                self.assertTrue(t in valid_types)
+
+                if t == 'option_string':
+                    # option_string type should have an options property that
+                    # describes the valid options
+                    self.assertTrue('options' in arg)
+                    # May be a list or dict because some option sets are self
+                    # explanatory and others need a description
```

Comment:
I think this comment is outdated now.

Reply:
whoops, yes it is! Though on [this thread](https://github.com/natcap/invest/pull/526#discussion_r648456636) I'm leaning back towards allowing `set`s... let me know what you think

Reply:
I think allowing sets is okay. Especially because the syntax is essentially the same, using `{ }` either way, and just including "values" where it makes sense. 👍

Reply:
Cool, I changed them to sets where before we had all the dictionary values be empty strings.